[libc++] Implement P2442R1 `std::views::slide`

shyeyian opened the pull request:
This PR implements P2442R1 in header `<__ranges/slide_view.h>`.

This is partial P2442 (as P2442 contains both `views::chunk` and `views::slide`).

Closes #171259

Commit:
Implement header <__ranges/slide.h>

Commit:
Add include info in related headers, e.g. <ranges> <version> and std-module.

Commit:
Add build info in related build files, e.g. CMakeLists.txt

Commit:
Append std tests for `views::slide`

Commit:
Append libcxx tests for `views::slide`

Commit:
Change the documents and release-notes on the current status

Commit:
Clang-format

shyeyian committed:
Generate meta info via xxx.py

Commit:
Fix CI build error: module `range.slide_view` did not imports exactly `iterator.prev`

Commit:
Fix CI build error: tests did not import std.memory or std.utility

anonymouspc commented:
@frederick-vs-ja Hello! May I request for a code-review here? This PR implemented `views::slide` corresponding to `views::chunk`.

Thanks, and Merry Christmas!

H-G-Hristov commented on libcxx/docs/ReleaseNotes/22.rst:
```diff
@@ -38,20 +38,18 @@ What's New in Libc++ 22.0.0?
 - P2944R3: Comparisons for ``reference_wrapper`` (`Github <https://llvm.org/PR105424>`__)
 - P3168R2: Give ``std::optional`` Range Support (`Github <https://llvm.org/PR105430>`__)
-- P3567R2: ``flat_meow`` Fixes (`Github <https://llvm.org/PR162022>`__)
+- P2442R1: P2442R1: Windowing range adaptors: ``views::chunk`` and ``views::slide`` (`Github <https://llvm.org/PR172948>`__) (Implemented ``views::slide`` only)
```
Please make sure to update this line if https://github.com/llvm/llvm-project/pull/171234 get's merge. We just need one entry.

frederick-vs-ja replied:
Also, we should avoid typing paper number (P2442R1) twice here (same for another PR).

anonymouspc replied:
Sounds good. I’ll keep it as “Implemented views::slide only” for now.
Once one of the PRs gets merged, I’ll update this line in the other PR so that we only keep a single entry.

H-G-Hristov commented on libcxx/include/__ranges/slide_view.h:
```diff
@@ -0,0 +1,396 @@
+      return __iterator<false>(ranges::begin(__base_), __n_);
+  }
+
+  [[nodiscard]] _LIBCPP_HIDDEN constexpr auto begin() const
```
Is this correct? `_LIBCPP_HIDDEN`

anonymouspc replied:
Thanks for pointing out this typo.

H-G-Hristov reviewed (commented):
Please make sure to link to the sub-task with "Closes #171259" by updating the top comment.

shyeyian committed:
Code Review: fix a typo in <__ranges/slide_view.h>

shyeyian committed:
Code Review: adjust release notes.

Note that we will further merge this commit with the one in #171234 in the future.

anonymouspc commented:
Thanks for your code review! The changes have been applied in the new commits.

-----

> Thank you for working on this. Can you fix the formatting errors.

Well, I believe the code has been formatted.
```
[anonymous@core clang]$ git branch
  main
  ranges_chunk
* ranges_slide
  sarif
[anonymous@core clang]$ git clang-format --diff $(git merge-base origin/main HEAD) HEAD
[anonymous@core clang]$ git-clang-format --diff origin/main HEAD --diff_from_common_commit
[anonymous@core clang]$ echo $?
0
[anonymous@core clang]$
``` 
It seems like the `github-actions` message is outdated (as it was last edited/updated a week ago).

shyeyian committed:
Add feature: assert if `slide_range` has a bad constructor arguments.

This commit also notes that `slide_range` should strictly have a `=default` constructor, according to [standard](https://eel.is/c++draft/range.slide#iterator)

shyeyian committed:
Add test: test if `slide_range` has a bad constructor argument.

See commit 1ddc459e43c8bbb8eb53e9c3081d9468fde69416.

Zingam commented:
> Thanks for your code review! The changes have been applied in the new commits.
> 
> > Thank you for working on this. Can you fix the formatting errors.
> 
> I reran `git clang-format` locally and it reports no differences:
> 
> ```
> [anonymous@core clang]$ git branch
>   main
>   ranges_chunk
> * ranges_slide
>   sarif
> [anonymous@core clang]$ git clang-format --diff $(git merge-base origin/main HEAD) HEAD
> [anonymous@core clang]$ git-clang-format --diff origin/main HEAD --diff_from_common_commit
> [anonymous@core clang]$ echo $?
> 0
> [anonymous@core clang]$
> ```
> 
> It looks like the `gitHub-actions` message may be outdated (as it was last updated a week ago). Please let me know if you still see a formatting issue on your side, I’m glad to update, thanks!

Have you tried rebasing? Please try it. There is a merge conflict too.

shyeyian committed:
Fix CI build: add `REQUIRES std-at-least-c++23` into `test/libcxx` tests

anonymouspc commented:
> Have you tried rebasing? Please try it. There is a merge conflict too.

I tried rebasing and then the `github-action` message get updated. Thank you!

> I pushed a commit to resolve conflicts. Fill free to force-push to eliminate it from the commit history.
> 
> Also, two PRs for P2441R1 will conflict with each other. The one merged later need to resolve the conflict again.

I've rebased them, and the commit history is now linear with no conflicts. Thanks!

shyeyian committed:
Fix CI: avoid adl `__to_unsigned_like`, and mark private member `using` as `_LIBCPP_NODEBUG`

ldionne committed:
Merge remote-tracking branch 'origin/main' into ranges_slide

# Conflicts:
#	libcxx/include/ranges

ldionne committed:
Fix release note
